#### This dataset is about rain in Australia. Here I will be trying to predict if it will rain tomorrow or not in specific dates.

In [11]:
import pandas as pd

df = pd.read_csv("weatherAUS.csv")
df.head()

Date Location  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  \
0  2008-12-01   Albury     13.4     22.9       0.6          NaN       NaN   
1  2008-12-02   Albury      7.4     25.1       0.0          NaN       NaN   
2  2008-12-03   Albury     12.9     25.7       0.0          NaN       NaN   
3  2008-12-04   Albury      9.2     28.0       0.0          NaN       NaN   
4  2008-12-05   Albury     17.5     32.3       1.0          NaN       NaN   

  WindGustDir  WindGustSpeed WindDir9am  ... Humidity9am  Humidity3pm  \
0           W           44.0          W  ...        71.0         22.0   
1         WNW           44.0        NNW  ...        44.0         25.0   
2         WSW           46.0          W  ...        38.0         30.0   
3          NE           24.0         SE  ...        45.0         16.0   
4           W           41.0        ENE  ...        82.0         33.0   

   Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  RainToday  \
0       1007.7       1007.1       8.0       NaN     16.9     21.8         No   
1       1010.6       1007.8       NaN       NaN     17.2     24.3         No   
2       1007.6       1008.7       NaN       2.0     21.0     23.2         No   
3       1017.6       1012.8       NaN       NaN     18.1     26.5         No   
4       1010.8       1006.0       7.0       8.0     17.8     29.7         No   

   RainTomorrow  
0            No  
1            No  
2            No  
3            No  
4            No  

[5 rows x 23 columns]

In [12]:
df.describe()

MinTemp        MaxTemp       Rainfall   Evaporation  \
count  143975.000000  144199.000000  142199.000000  82670.000000   
mean       12.194034      23.221348       2.360918      5.468232   
std         6.398495       7.119049       8.478060      4.193704   
min        -8.500000      -4.800000       0.000000      0.000000   
25%         7.600000      17.900000       0.000000      2.600000   
50%        12.000000      22.600000       0.000000      4.800000   
75%        16.900000      28.200000       0.800000      7.400000   
max        33.900000      48.100000     371.000000    145.000000   

           Sunshine  WindGustSpeed   WindSpeed9am   WindSpeed3pm  \
count  75625.000000  135197.000000  143693.000000  142398.000000   
mean       7.611178      40.035230      14.043426      18.662657   
std        3.785483      13.607062       8.915375       8.809800   
min        0.000000       6.000000       0.000000       0.000000   
25%        4.800000      31.000000       7.000000      13.000000   
50%        8.400000      39.000000      13.000000      19.000000   
75%       10.600000      48.000000      19.000000      24.000000   
max       14.500000     135.000000     130.000000      87.000000   

         Humidity9am    Humidity3pm   Pressure9am    Pressure3pm  \
count  142806.000000  140953.000000  130395.00000  130432.000000   
mean       68.880831      51.539116    1017.64994    1015.255889   
std        19.029164      20.795902       7.10653       7.037414   
min         0.000000       0.000000     980.50000     977.100000   
25%        57.000000      37.000000    1012.90000    1010.400000   
50%        70.000000      52.000000    1017.60000    1015.200000   
75%        83.000000      66.000000    1022.40000    1020.000000   
max       100.000000     100.000000    1041.00000    1039.600000   

           Cloud9am      Cloud3pm        Temp9am       Temp3pm  
count  89572.000000  86102.000000  143693.000000  141851.00000  
mean       4.447461      4.509930      16.990631      21.68339  
std        2.887159      2.720357       6.488753       6.93665  
min        0.000000      0.000000      -7.200000      -5.40000  
25%        1.000000      2.000000      12.300000      16.60000  
50%        5.000000      5.000000      16.700000      21.10000  
75%        7.000000      7.000000      21.600000      26.40000  
max        9.000000      9.000000      40.200000      46.70000

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 145460 entries, 0 to 145459
Data columns (total 23 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   Date           145460 non-null  object 
 1   Location       145460 non-null  object 
 2   MinTemp        143975 non-null  float64
 3   MaxTemp        144199 non-null  float64
 4   Rainfall       142199 non-null  float64
 5   Evaporation    82670 non-null   float64
 6   Sunshine       75625 non-null   float64
 7   WindGustDir    135134 non-null  object 
 8   WindGustSpeed  135197 non-null  float64
 9   WindDir9am     134894 non-null  object 
 10  WindDir3pm     141232 non-null  object 
 11  WindSpeed9am   143693 non-null  float64
 12  WindSpeed3pm   142398 non-null  float64
 13  Humidity9am    142806 non-null  float64
 14  Humidity3pm    140953 non-null  float64
 15  Pressure9am    130395 non-null  float64
 16  Pressure3pm    130432 non-null  float64
 17  Cloud9am       89572 non-null

#### Let's count how many null values we have:

In [14]:
nulls = df.isnull().sum().sort_values()
nulls

Date                 0
Location             0
MaxTemp           1261
MinTemp           1485
Temp9am           1767
WindSpeed9am      1767
Humidity9am       2654
WindSpeed3pm      3062
Rainfall          3261
RainToday         3261
RainTomorrow      3267
Temp3pm           3609
WindDir3pm        4228
Humidity3pm       4507
WindGustSpeed    10263
WindGustDir      10326
WindDir9am       10566
Pressure3pm      15028
Pressure9am      15065
Cloud9am         55888
Cloud3pm         59358
Evaporation      62790
Sunshine         69835
dtype: int64

In [15]:
total = len(df)
null_percent = (nulls/total).sort_values()
null_percent

Date             0.000000
Location         0.000000
MaxTemp          0.008669
MinTemp          0.010209
Temp9am          0.012148
WindSpeed9am     0.012148
Humidity9am      0.018246
WindSpeed3pm     0.021050
Rainfall         0.022419
RainToday        0.022419
RainTomorrow     0.022460
Temp3pm          0.024811
WindDir3pm       0.029066
Humidity3pm      0.030984
WindGustSpeed    0.070555
WindGustDir      0.070989
WindDir9am       0.072639
Pressure3pm      0.103314
Pressure9am      0.103568
Cloud9am         0.384216
Cloud3pm         0.408071
Evaporation      0.431665
Sunshine         0.480098
dtype: float64

In [16]:
most_nulls = pd.concat([nulls, null_percent], axis=1, keys=["Total", "Percent"])
most_nulls

Total   Percent
Date               0  0.000000
Location           0  0.000000
MaxTemp         1261  0.008669
MinTemp         1485  0.010209
Temp9am         1767  0.012148
WindSpeed9am    1767  0.012148
Humidity9am     2654  0.018246
WindSpeed3pm    3062  0.021050
Rainfall        3261  0.022419
RainToday       3261  0.022419
RainTomorrow    3267  0.022460
Temp3pm         3609  0.024811
WindDir3pm      4228  0.029066
Humidity3pm     4507  0.030984
WindGustSpeed  10263  0.070555
WindGustDir    10326  0.070989
WindDir9am     10566  0.072639
Pressure3pm    15028  0.103314
Pressure9am    15065  0.103568
Cloud9am       55888  0.384216
Cloud3pm       59358  0.408071
Evaporation    62790  0.431665
Sunshine       69835  0.480098

#### Those last 4 values are too empty. I have decided to get rid of them.

In [17]:
dropList = list(most_nulls[-4:].index)
df.drop(dropList, axis=1, inplace=True)
dropList

['Cloud9am', 'Cloud3pm', 'Evaporation', 'Sunshine']

In [18]:
print(len(df))
print(len(df.dropna()))
df = df.dropna()

145460
112925


#### Now it is time to convert all the "object" values into numbers. Later I would use OHE.

In [19]:
len(df["Location"].unique())

44

In [20]:
new_values = {}
for i in range(len(df["Location"].unique())):
    new_values.update({df["Location"].unique()[i]:i})

new_values

{'Albury': 0,
 'BadgerysCreek': 1,
 'Cobar': 2,
 'CoffsHarbour': 3,
 'Moree': 4,
 'NorahHead': 5,
 'NorfolkIsland': 6,
 'Richmond': 7,
 'Sydney': 8,
 'SydneyAirport': 9,
 'WaggaWagga': 10,
 'Williamtown': 11,
 'Wollongong': 12,
 'Canberra': 13,
 'Tuggeranong': 14,
 'Ballarat': 15,
 'Bendigo': 16,
 'Sale': 17,
 'MelbourneAirport': 18,
 'Melbourne': 19,
 'Mildura': 20,
 'Nhil': 21,
 'Portland': 22,
 'Watsonia': 23,
 'Dartmoor': 24,
 'Brisbane': 25,
 'Cairns': 26,
 'GoldCoast': 27,
 'Townsville': 28,
 'Adelaide': 29,
 'MountGambier': 30,
 'Nuriootpa': 31,
 'Woomera': 32,
 'Witchcliffe': 33,
 'PearceRAAF': 34,
 'PerthAirport': 35,
 'Perth': 36,
 'Walpole': 37,
 'Hobart': 38,
 'Launceston': 39,
 'AliceSprings': 40,
 'Darwin': 41,
 'Katherine': 42,
 'Uluru': 43}

In [21]:
def values_to_nums(value):
    return new_values[value]

df["Location"] = df["Location"].apply(values_to_nums)
df["Location"].unique()

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
       34, 35, 36, 37, 38, 39, 40, 41, 42, 43], dtype=int64)

In [22]:
len(df["Date"].unique())

3417

In [23]:
df.drop("Date", axis=1, inplace=True)
df.head()

Location  MinTemp  MaxTemp  Rainfall WindGustDir  WindGustSpeed WindDir9am  \
0         0     13.4     22.9       0.6           W           44.0          W   
1         0      7.4     25.1       0.0         WNW           44.0        NNW   
2         0     12.9     25.7       0.0         WSW           46.0          W   
3         0      9.2     28.0       0.0          NE           24.0         SE   
4         0     17.5     32.3       1.0           W           41.0        ENE   

  WindDir3pm  WindSpeed9am  WindSpeed3pm  Humidity9am  Humidity3pm  \
0        WNW          20.0          24.0         71.0         22.0   
1        WSW           4.0          22.0         44.0         25.0   
2        WSW          19.0          26.0         38.0         30.0   
3          E          11.0           9.0         45.0         16.0   
4         NW           7.0          20.0         82.0         33.0   

   Pressure9am  Pressure3pm  Temp9am  Temp3pm RainToday RainTomorrow  
0       1007.7       1007.1     16.9     21.8        No           No  
1       1010.6       1007.8     17.2     24.3        No           No  
2       1007.6       1008.7     21.0     23.2        No           No  
3       1017.6       1012.8     18.1     26.5        No           No  
4       1010.8       1006.0     17.8     29.7        No           No

In [24]:
df.select_dtypes(object)

WindGustDir WindDir9am WindDir3pm RainToday RainTomorrow
0                W          W        WNW        No           No
1              WNW        NNW        WSW        No           No
2              WSW          W        WSW        No           No
3               NE         SE          E        No           No
4                W        ENE         NW        No           No
...            ...        ...        ...       ...          ...
145454           E        ESE          E        No           No
145455           E         SE        ENE        No           No
145456         NNW         SE          N        No           No
145457           N         SE        WNW        No           No
145458          SE        SSE          N        No           No

[112925 rows x 5 columns]

In [25]:
print(df["WindGustDir"].dropna().unique())
print(df["WindDir9am"].dropna().unique())
print(df["WindDir3pm"].dropna().unique())

['W' 'WNW' 'WSW' 'NE' 'NNW' 'N' 'NNE' 'SW' 'ENE' 'SSE' 'S' 'NW' 'SE' 'ESE'
 'E' 'SSW']
['W' 'NNW' 'SE' 'ENE' 'SW' 'SSE' 'S' 'NE' 'SSW' 'N' 'WSW' 'ESE' 'E' 'NW'
 'WNW' 'NNE']
['WNW' 'WSW' 'E' 'NW' 'W' 'SSE' 'ESE' 'ENE' 'NNW' 'SSW' 'SW' 'SE' 'N' 'S'
 'NNE' 'NE']


In [26]:
wind_values = {}
for i in range(len(df["WindGustDir"].dropna().unique())):
    wind_values.update({df["WindGustDir"].dropna().unique()[i]:i})
    
wind_values

{'W': 0,
 'WNW': 1,
 'WSW': 2,
 'NE': 3,
 'NNW': 4,
 'N': 5,
 'NNE': 6,
 'SW': 7,
 'ENE': 8,
 'SSE': 9,
 'S': 10,
 'NW': 11,
 'SE': 12,
 'ESE': 13,
 'E': 14,
 'SSW': 15}

In [27]:
def wind_to_nums(value):
    return wind_values[value]

df["WindGustDir"] = df["WindGustDir"].apply(wind_to_nums)
df["WindDir9am"] = df["WindDir9am"].apply(wind_to_nums)
df["WindDir3pm"] = df["WindDir3pm"].apply(wind_to_nums)
df.head()

Location  MinTemp  MaxTemp  Rainfall  WindGustDir  WindGustSpeed  \
0         0     13.4     22.9       0.6            0           44.0   
1         0      7.4     25.1       0.0            1           44.0   
2         0     12.9     25.7       0.0            2           46.0   
3         0      9.2     28.0       0.0            3           24.0   
4         0     17.5     32.3       1.0            0           41.0   

   WindDir9am  WindDir3pm  WindSpeed9am  WindSpeed3pm  Humidity9am  \
0           0           1          20.0          24.0         71.0   
1           4           2           4.0          22.0         44.0   
2           0           2          19.0          26.0         38.0   
3          12          14          11.0           9.0         45.0   
4           8          11           7.0          20.0         82.0   

   Humidity3pm  Pressure9am  Pressure3pm  Temp9am  Temp3pm RainToday  \
0         22.0       1007.7       1007.1     16.9     21.8        No   
1         25.0       1010.6       1007.8     17.2     24.3        No   
2         30.0       1007.6       1008.7     21.0     23.2        No   
3         16.0       1017.6       1012.8     18.1     26.5        No   
4         33.0       1010.8       1006.0     17.8     29.7        No   

  RainTomorrow  
0           No  
1           No  
2           No  
3           No  
4           No

In [28]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 112925 entries, 0 to 145458
Data columns (total 18 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   Location       112925 non-null  int64  
 1   MinTemp        112925 non-null  float64
 2   MaxTemp        112925 non-null  float64
 3   Rainfall       112925 non-null  float64
 4   WindGustDir    112925 non-null  int64  
 5   WindGustSpeed  112925 non-null  float64
 6   WindDir9am     112925 non-null  int64  
 7   WindDir3pm     112925 non-null  int64  
 8   WindSpeed9am   112925 non-null  float64
 9   WindSpeed3pm   112925 non-null  float64
 10  Humidity9am    112925 non-null  float64
 11  Humidity3pm    112925 non-null  float64
 12  Pressure9am    112925 non-null  float64
 13  Pressure3pm    112925 non-null  float64
 14  Temp9am        112925 non-null  float64
 15  Temp3pm        112925 non-null  float64
 16  RainToday      112925 non-null  object 
 17  RainTomorrow   112925 non-nul

In [29]:
def bool_to_nums(value):
    if "no" in value.lower():
        return 0
    if "yes" in value.lower():
        return 1
    
df["RainToday"] = df["RainToday"].apply(bool_to_nums)
df["RainTomorrow"] = df["RainTomorrow"].apply(bool_to_nums)
df.head()

Location  MinTemp  MaxTemp  Rainfall  WindGustDir  WindGustSpeed  \
0         0     13.4     22.9       0.6            0           44.0   
1         0      7.4     25.1       0.0            1           44.0   
2         0     12.9     25.7       0.0            2           46.0   
3         0      9.2     28.0       0.0            3           24.0   
4         0     17.5     32.3       1.0            0           41.0   

   WindDir9am  WindDir3pm  WindSpeed9am  WindSpeed3pm  Humidity9am  \
0           0           1          20.0          24.0         71.0   
1           4           2           4.0          22.0         44.0   
2           0           2          19.0          26.0         38.0   
3          12          14          11.0           9.0         45.0   
4           8          11           7.0          20.0         82.0   

   Humidity3pm  Pressure9am  Pressure3pm  Temp9am  Temp3pm  RainToday  \
0         22.0       1007.7       1007.1     16.9     21.8          0   
1         25.0       1010.6       1007.8     17.2     24.3          0   
2         30.0       1007.6       1008.7     21.0     23.2          0   
3         16.0       1017.6       1012.8     18.1     26.5          0   
4         33.0       1010.8       1006.0     17.8     29.7          0   

   RainTomorrow  
0             0  
1             0  
2             0  
3             0  
4             0

#### Now that the data is sorted out, it is time to start with the KNN model.

In [30]:
y = df["RainTomorrow"]
X = df.drop(["RainTomorrow"], axis=1)

In [31]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.50)

In [32]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
scaler.fit(X_train)
X_train = scaler.transform(X_train)
X_test = scaler.transform(X_test)
X_train

array([[-1.22526259,  0.08163531, -0.85446609, ..., -0.37344419,
        -0.73766709, -0.53733945],
       [ 0.39469314,  1.61588211,  1.32338064, ...,  1.63981722,
         1.50104441, -0.53733945],
       [-0.09129358,  0.32136137, -0.43895586, ..., -0.29480117,
        -0.38649666, -0.53733945],
       ...,
       [ 0.55668871,  0.17752574,  0.83623072, ...,  0.96348722,
         0.81333565, -0.53733945],
       [-0.98226923,  1.08848477,  0.2058014 , ...,  0.86911559,
         0.2426837 , -0.53733945],
       [ 1.60965994,  1.07250303,  1.43800416, ...,  1.79710327,
         1.67662962, -0.53733945]])

In [33]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

check = []
preds = []
for k in range(1, 30):
    classifier = KNeighborsClassifier(n_neighbors = k)
    classifier.fit(X_train, y_train)
    y_pred = classifier.predict(X_test)
    preds.append(y_pred)
    check.append(accuracy_score(y_test,y_pred))
    
print(check.index(max(check)))
y_pred = preds[check.index(max(check))]

# classifier = KNeighborsClassifier(n_neighbors = 28)
# classifier.fit(X_train, y_train)
# y_pred = classifier.predict(X_test)

28


#### The best number of K nearest neighbors I have found is 28, and the accuracy is 84.3%.

In [34]:
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score
result = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:")
print(result)
result1 = classification_report(y_test, y_pred)
print("Classification Report:",)
print (result1)
result2 = accuracy_score(y_test,y_pred)
print("Accuracy:",result2)

Confusion Matrix:
[[42169  1769]
 [ 6976  5549]]
Classification Report:
              precision    recall  f1-score   support

           0       0.86      0.96      0.91     43938
           1       0.76      0.44      0.56     12525

    accuracy                           0.85     56463
   macro avg       0.81      0.70      0.73     56463
weighted avg       0.84      0.85      0.83     56463

Accuracy: 0.8451198129748685


#### Let's try this time without the Location.

In [35]:
y = df["RainTomorrow"]
X = df.drop(["RainTomorrow","Location"], axis=1)

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.50)

from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
scaler.fit(X_train)
X_train = scaler.transform(X_train)
X_test = scaler.transform(X_test)

check = []
preds = []
for k in range(1, 30):
    classifier = KNeighborsClassifier(n_neighbors = k)
    classifier.fit(X_train, y_train)
    y_pred = classifier.predict(X_test)
    preds.append(y_pred)
    check.append(accuracy_score(y_test,y_pred))

print(check.index(max(check)))
y_pred = preds[check.index(max(check))]

# classifier = KNeighborsClassifier(n_neighbors = 28)
# classifier.fit(X_train, y_train)
# y_pred = classifier.predict(X_test)

from sklearn.metrics import classification_report, confusion_matrix, accuracy_score
result = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:")
print(result)
result1 = classification_report(y_test, y_pred)
print("Classification Report:",)
print (result1)
result2 = accuracy_score(y_test,y_pred)
print("Accuracy:",result2)

28
Confusion Matrix:
[[42073  1858]
 [ 6959  5573]]
Classification Report:
              precision    recall  f1-score   support

           0       0.86      0.96      0.91     43931
           1       0.75      0.44      0.56     12532

    accuracy                           0.84     56463
   macro avg       0.80      0.70      0.73     56463
weighted avg       0.83      0.84      0.83     56463

Accuracy: 0.8438446416237182


#### Let's use OHE now for the Winds.

In [36]:
df = pd.read_csv("weatherAUS.csv")
df.drop(dropList, axis=1, inplace=True)
df.drop("Date", axis=1, inplace=True)
df.drop("Location", axis=1, inplace=True)

In [37]:
from sklearn.preprocessing import OneHotEncoder
enc = pd.get_dummies(data=df, columns=['WindGustDir','WindDir9am','WindDir3pm'])
enc.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 145460 entries, 0 to 145459
Data columns (total 62 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   MinTemp          143975 non-null  float64
 1   MaxTemp          144199 non-null  float64
 2   Rainfall         142199 non-null  float64
 3   WindGustSpeed    135197 non-null  float64
 4   WindSpeed9am     143693 non-null  float64
 5   WindSpeed3pm     142398 non-null  float64
 6   Humidity9am      142806 non-null  float64
 7   Humidity3pm      140953 non-null  float64
 8   Pressure9am      130395 non-null  float64
 9   Pressure3pm      130432 non-null  float64
 10  Temp9am          143693 non-null  float64
 11  Temp3pm          141851 non-null  float64
 12  RainToday        142199 non-null  object 
 13  RainTomorrow     142193 non-null  object 
 14  WindGustDir_E    145460 non-null  uint8  
 15  WindGustDir_ENE  145460 non-null  uint8  
 16  WindGustDir_ESE  145460 non-null  uint

#### Trying a different method of converting RainToday and RainTomorrow to numbers. At the end it was the same as earlier.

In [38]:
from sklearn import preprocessing
from numpy import array

enc['RainToday'] = df['RainToday'].astype(str)
enc['RainTomorrow'] = df['RainTomorrow'].astype(str)

lb = preprocessing.LabelBinarizer()

enc['RainToday'] = lb.fit_transform(enc['RainToday'])
enc['RainTomorrow'] = lb.fit_transform(enc['RainTomorrow'])

enc = enc.dropna()

enc[["RainToday","RainTomorrow"]].head(10)

RainToday  RainTomorrow
0          1             1
1          1             1
2          1             1
3          1             1
4          1             1
5          1             1
6          1             1
7          1             1
8          1             0
9          0             1

In [39]:
y = enc['RainTomorrow']
X = enc.drop(['RainTomorrow'], axis=1)

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.50)

from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
scaler.fit(X_train)
X_train = scaler.transform(X_train)
X_test = scaler.transform(X_test)

from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

check = []
preds = []
for k in range(1, 30):
    classifier = KNeighborsClassifier(n_neighbors = k)
    classifier.fit(X_train, y_train)
    y_pred = classifier.predict(X_test)
    preds.append(y_pred)
    check.append(accuracy_score(y_test,y_pred))

print(check.index(max(check)))
y_pred = preds[check.index(max(check))]

# classifier = KNeighborsClassifier(n_neighbors = 17)
# classifier.fit(X_train, y_train)
# y_pred = classifier.predict(X_test)

from sklearn.metrics import classification_report, confusion_matrix, accuracy_score
result = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:")
print(result)
result1 = classification_report(y_test, y_pred)
print("Classification Report:",)
print (result1)
result2 = accuracy_score(y_test,y_pred)
print("Accuracy:",result2)

#### The results were slightly higher. accuracy of 84.4%. Lets try with Logistic regression.

In [ ]:
from sklearn.linear_model import LogisticRegression
model = LogisticRegression()
model.fit(X_train, y_train)

In [ ]:
y_pred = model.predict(X_test)
y_pred

In [ ]:
model.score(X_test,y_test)

In [ ]:
model.predict_proba(X_test)

#### Accuracy of 84.5%.